src/simple_lottery/init.rs

```diff
@@ -0,0 +1,94 @@
+//! Parameter derivation for simple lottery.
+
+use super::params::Params;
+use super::setup::Setup;
+
+use std::f64::consts::LN_2;
+
+/// Calculates Setup parameters.
+pub fn make_setup(params: &Params) -> Setup {
+    let ratio = params.set_size as f64 / params.lower_bound as f64;
+
+    // Execute binary search for the optimal value of `rs`.
+
+    let mut l: f64 = 0.0;
+    let mut r: f64 = ratio;
+
+    loop {
+        let m = (l + r) / 2.0;
+        let rs = m;
+        let rc = ratio / rs;
+
+        let lhs = params.soundness_param / (rs.ln() - 1.0 + 1.0 / rs);
+        let rhs = params.completeness_param / (rc - 1.0 - rc.ln());
+
```

Comment:
Since `left` and `right` are floating point numbers, don't you want to check some sort of convergence (e.g., if `right - left < EPS`, you break the loop)?

Reply:
I would not say that the goal of doing this is to optimize the code though but to follow the logic of what we want to do: we want to find `rs` such that the u given by each bounds are equal.

Reply:
Let's go a step backward to see what we are trying to achieve. 

The code here has a clear goal: finding the minimal `u`. We do so by doing a dichotomic search on `rs`. What are the conditions when iterating on `rs`? We have 3: `rs` and `rc` must be stricter than  - this is already implicitely handled by the bounds of the dichotomic search- and we want to converge the soundness and completeness bounds, that is `|(lhs * ln2).ceil() - (rhs * ln2).ceil()| = 0` .

Now, this repository is supposed to be a reference implementation, you say yourself in the design document. Hence, we are aiming at being clear about what each piece of code is doing. Do we think that we should have undocumented code that goes straight for the solution without explaining some intricacies, not only about the algorithm but also how it was implemented (for instance the variable names, why looping as done here is fine and the expected number of iterations (your talk about f64 precision))? Or should we make an effort to be clear by having meaningful and coherent variable names, checks and tests? I lean towards the latter.

Most of the review comments, and more globally general decisions, so far have aimed at that goal. There has of course also been here and there some changes and optimizations asked, I won't deny that. But if you look at the history of this repository, you will see that we have already made some along the way that were deemed reasonable at some point in time (for instance, reusing values in the `centralized_telescope` setup) and so it is coherent to continue in that fashion. It is in the developer's, reviewer's best interest to keep a similar coding style afterall.

 To go back to the bound, `|(lhs * ln2).ceil() - (rhs * ln2).ceil()| = 0`. 
> Aren't I talking about optimizing the code here? 

I am, it is an optimization, and that is fine. It is also not my only point.
    
> Which of the goals does the current code not solve?

The code is not the clearest. We should strive for better variable names, more documentation, and  specifying what we are trying to do. The paper moreover can be hard to understand, you highlighted before some of its weaknesses and people have been confused about parts of it (the role of `nf` and `np` for instance as confounded more than one person, I also personally discussed with Pyrrhos more than once about part of the paper). We have an opportunity here to improve on it, let's do it.

Reply:
What I mean is, mathematically `lhs` and `rhs` exist, but given that we use floating-point numbers with fixed precision, there might not a representation that satisfy the formula.

What comments would you like to add?